be/src/exprs/in_const_predicate.hpp

```diff
@@ -395,6 +399,120 @@ class VectorizedInConstPredicate final : public Predicate {
     std::vector<ColumnPtr> _string_values;
 };
 
+class VectorizedInConstPredicateGeneric final : public Predicate {
+public:
+    VectorizedInConstPredicateGeneric(const TExprNode& node)
+            : Predicate(node), _is_not_in(node.in_predicate.is_not_in) {}
+
+    VectorizedInConstPredicateGeneric(const VectorizedInConstPredicateGeneric& other)
+            : Predicate(other), _is_not_in(other._is_not_in) {}
+
+    ~VectorizedInConstPredicateGeneric() override = default;
+
+    Expr* clone(ObjectPool* pool) const override { return pool->add(new VectorizedInConstPredicateGeneric(*this)); }
+
+    Status open(RuntimeState* state, ExprContext* context, FunctionContext::FunctionStateScope scope) override {
+        RETURN_IF_ERROR(Expr::open(state, context, scope));
+        _const_input.resize(_children.size());
+        for (auto i = 0; i < _children.size(); ++i) {
+            if (dynamic_cast<VectorizedLiteral*>(_children[i])) {
+                ASSIGN_OR_RETURN(_const_input[i], _children[i]->evaluate_checked(context, nullptr));
```

Comment:
the children of `in` must be constant, I think you can add DCHECK on `expr.is_const`,  and don't need complex handle

Reply:
seems `Expr::is_constant()` is not overrided by any expression.

Reply:
as currently rewritting non-const parameters of in predict to OR  expressions results in slower performance, we could optimize it by allow in predict to support non-const parameteres, so it is necessary to keep this generial process.

Reply:
> seems `Expr::is_constant()` is not overrided by some expression, and this interface is not really used.

1. only a few functions doesn't support `is_constant`, such as `random`, don't worry.
2. as you say,  I think rewrite contant expression on `IN` predicate is unnecessary, but it's only need to modify FE. you can think about it, execut a constant expression once more is faster than execute it as a variable column in your template

Reply:
BTW, support variable column on `IN` is not necessarily faster than using OR